python/Ganga/GPIDev/Base/Objects.py

```diff
@@ -785,6 +810,88 @@ def __construct__(self, args):
             from Ganga.GPIDev.Base.Proxy import TypeMismatchError
             raise TypeMismatchError("Constructor expected one or zero non-keyword arguments, got %i" % len(args))
 
+    @staticmethod
+    def _parentString(parent):
+        registry = parent._getRegistry()
+
+        if registry is not None:
+            parent_id = registry.find(parent)
+            parent_str = _getName(registry) + "_" + str(parent_id)
+        else:
+            parent_id = id(parent)
+            parent_str = "UnknownReg_" + str(parent_id)
+
+        return parent_str
+
+    @staticmethod
+    def _acquireChildLock(child_name, parent):
+        child_path = GangaObject._getChildPathStr(parent)
+
+        whole_path = child_path + "_" + child_name
+
+        global childLockDict
+        if whole_path not in childLockDict:
+            import threading
+            childLockDict[whole_path] = threading.Lock()
+
+        childLockDict[whole_path].acquire()
+
+    @staticmethod
+    def _releaseChildLock(child_name, parent):
+        child_path = GangaObject._getChildPathStr(parent)
+
+        whole_path = child_path + "_" + child_name
+
+        global childLockDict
+        if whole_path in childLockDict:
+            childLockDict[whole_path].release()
+            return
+
+        raise GangaException("Cannot release Lock: %s" % whole_path)
+
+    @staticmethod
+    def _getChildPathStr(child):
+        child_str = _getName(child)
+        parent = child._getParent()
+
+        while parent is not None:
+            child_str += "_" + _getName(parent)
+            parent = parent._getParent()
+
+        return child_str
+
+    @staticmethod
+    def _acquireParentLock(child):
+        ## Top level parent
+        parent = child._getRoot()
+
+        parent_str = GangaObject._parentString(parent)
+
+        if parent is None:
+            raise GangaException("Cannot Lock an object without a parent!")
+
+        global parentLockDict
+        if parent_str not in parentLockDict:
+            import threading
+            parentLockDict[parent_str] = threading.Lock()
+
+        parentLockDict[parent_str].acquire()
+
+    @staticmethod
+    def _releaseParentLock(child):
+        ## Top level parent
+        parent = child._getRoot()
+
+        parent_str = GangaObject._parentString(parent)
+
+        if parent is not None and parent_str in parentLockDict:
+
+            parentLockDict[parent_str].release()
+            return
+
+        raise GangaException("Cannot release an object if it has no parent associated with it: %s" % parent_str)
+
+
```

Comment:
Instead of all these functions, could we not store the lock inside the root object for each tree rather than in a global dictionary? so
```python
@staticmethod
def _releaseParentLock(child):
    ## Top level parent
    child._getRoot().lock.release()
```
except with a little checking if there no parent?